core/src/main/scala/org/apache/spark/scheduler/MapStatus.scala

```diff
@@ -28,16 +28,23 @@ import org.apache.spark.internal.config
 import org.apache.spark.storage.BlockManagerId
 import org.apache.spark.util.Utils
 
+trait Location extends Externalizable {
```

Comment:
`Externalizable` would require the implementers to provide a zero-parameter constructor. Do you think it's acceptable or have any better idea? @hiboyang @attilapiros

Reply:
Because the `BlockManagerId` already used `Externalizable` and we want to make the `BlockManagerId` as a default implementation of `Location` for Spark.

A possible way to let `Location` extends `Serializable` is to let `BlockManagerId` extends `Serializable` directly too. But we need to figure out whether serializers could handle `topologyInfo_: Option[String]` properly as `topologyInfo_` now is handled manually in `readExternal/writeExternal`:

https://github.com/apache/spark/blob/8218b488035049434271dc9e3bd5af45ffadf0fd/core/src/main/scala/org/apache/spark/storage/BlockManagerId.scala#L71-L86

Reply:
To clarify: `Externalizable` takes precedence over `Serializable`.
We can continue to use `Externalizable` for `BlockManagerId` while not requiring other implementations to necessarily require it.